File: iosApp/iosApp/Modules/Degen/Cult/CultProposals/CultProposalsViewModel.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import UIKit

enum CultProposalsViewModel {
    case loading
    case loaded(sections: [Section], selectedSectionType: Section.`Type`)
    case error(error: AppsViewModel.Error)
}

extension CultProposalsViewModel {
    
    struct Section {
        
        let title: String
        let type: `Type`
        let items: [Item]
        
        enum `Type` {
            case pending
            case closed
        }
    }

    struct Item {
        
        let id: String
        let title: String
        let approved: Vote
        let rejected: Vote
        let approveButtonTitle: String
        let rejectButtonTitle: String
        let endDate: Date
        
        struct Vote {
            
            let name: String
            let value: Double
            let total: Double
            let type: `Type`
            
            enum `Type` {
                
                case approved
                case rejected
            }
        }
    }
}

extension CultProposalsViewModel {

    struct Error {
        
        let title: String
        let body: String
        let actions: [String]
    }
}

extension CultProposalsViewModel {
    
    var title: String {
        Localized("cult.proposals.title")
    }
    
    var titleIcon: Data {
        
        UIImage(named: "degen-cult-icon")!.pngData()!
    }

    var sections: [CultProposalsViewModel.Section] {
        
        switch self {
        case let .loaded(sections, _):
            return sections
        default:
            return []
        }
    }
    
    var selectedSectionType: CultProposalsViewModel.Section.`Type` {
        
        switch self {
        case let .loaded(_, type):
            return type
        default:
            return .pending
        }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/degenCultProposals/CultProposalsViewModel.kt
package com.sonsofcrypto.web3walletcore.modules.degenCultProposals

import com.sonsofcrypto.web3walletcore.common.viewModels.ErrorViewModel

sealed class CultProposalsViewModel {
    object Loading: CultProposalsViewModel()
    data class Loaded(val sections: List<Section>): CultProposalsViewModel()
    data class Error(val error: ErrorViewModel): CultProposalsViewModel()

    sealed class Section() {
        data class Pending(val title: String, val items: List<Item>, val footer: Footer): Section()
        data class Closed(val title: String, val items: List<Item>, val footer: Footer): Section()
    }

    data class Item(
        val id: String,
        val title: String,
        val approved: Vote,
        val rejected: Vote,
        val approveButtonTitle: String,
        val rejectButtonTitle: String,
        val endDate: Double,
        val stateName: String,
    )

    data class Vote(
        val name: String,
        val value: Double,
        val total : Double,
    )

    data class Footer(
        val imageName: String,
        val text: String,
    )
}
